jwt.js

```diff
@@ -1,61 +1,114 @@
-'use strict'
-
 const fp = require('fastify-plugin')
 const JWT = require('jsonwebtoken')
-const assert = require('assert')
+const steed = require('steed')()
+
+function wrapStaticSecretInCallback (secret) {
+  return function (_, __, cb) {
+    return cb(null, secret)
+  }
+}
 
-function fastifyJWT (fastify, opts, next) {
-  if (!opts.secret) {
+function fastifyJwt (fastify, options, next) {
+  if (!options.secret) {
     return next(new Error('missing secret'))
   }
 
-  const secret = opts.secret
+  let secretCallback = options.secret
+
+  if (typeof secretCallback !== 'function') { secretCallback = wrapStaticSecretInCallback(secretCallback) }
+
+  let _requestProperty = options.userProperty || options.requestProperty || 'user'
+  let _resultProperty = options.resultProperty
+  const credentialsRequired = typeof options.credentialsRequired === 'undefined' ? true : options.credentialsRequired
 
   fastify.decorate('jwt', {
     sign: sign,
     verify: verify,
-    decode: decode,
-    secret: secret
+    secretCallback: secretCallback
   })
 
   next()
 
-  function sign (payload, options, callback) {
-    assert(payload, 'missing payload')
-    options = options || {}
-    if (typeof options === 'function') {
-      callback = options
-      options = {}
-    }
+  function sign (request, reply, next) {
+    let {secret, resultProperty, ...rest} = options
+    steed.waterfall([
+      function getSecret (callback) {
+        secretCallback({}, {}, callback)
+      },
+      function sign (secret, callback) {
+        JWT.sign(request.body, secret, rest, function (err, token) {
+          if (err) {
+            callback(err)
+          } else {
+            callback(null, token)
+          }
+        })
+      }
+    ], function (err, result) {
+      if (_resultProperty) {
+        reply[_resultProperty] = result
+      } else {
+        request[_requestProperty] = result
+      }
```

Comment:
I'm 👎  in doing this. Why do you need to modify these?

Reply:
This was abstracted from `express-jwt`. If you needed access to the result of either sign or verify on your request/reply then this lets you specify where you can find it.
```js
fastify.register(jwt, {
  secret: 'shh',
  resultProperty: 'foo'
})
fastify.get(
  '/api',
  {
    beforeHandler: [
      function (req, res, done) {
        fastify.jwt.verify(req, res, done)
      },
      function (req, res, done) {
        console.log(res.foo) // <- the result of jwt.verify
      }
    ]
  }, 
  (request, reply) => reply.send({ fuzz: 'buzz' })
)
```

This is the first time I've ever really handled middleware so you tell me if this is bad practice or not.

Reply:
I think that's bad practice. If we need to add things in this way we should do:

```js
request.verifyJwt(cb) // request.verifyJwt().then()
```

I'm starting to think that this PR should add new decorators to `Request`  and `Reply` and __keep and use__ the sign/verify methods that were attached to the main instance.

Reply:
Interesting, how would this be implemented in a beforeHandler method? Or would we execute these functions directly in the route handler?